File: 0075-sort-colors/0075-sort-colors.py
class Solution:
    def sortColors(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        left, idx = 0, 0
        right = len(nums) -1 
        while idx <= right:
            # 3 rules
            if nums[idx] == 0:
                nums[left], nums[idx] = nums[idx], nums[left]
                idx += 1
                left += 1
            elif nums[idx] == 1:
                idx += 1
            else:                
                nums[right], nums[idx] = nums[idx], nums[right]
                right -= 1

File: 0075-sort-colors/0075-sort-colors.java
class Solution {
    public void sortColors(int[] nums) {
        int left = 0, right = nums.length-1;
        int zero = 0;
        while(left <= right){
            if(nums[left] == 0){
                int temp = nums[zero];
                nums[zero] = nums[left];
                nums[left] = temp;
                zero++;
                left++;
            }else if(nums[left] == 2){
                //swap the number with right
                int temp = nums[left];
                nums[left] = nums[right];
                nums[right] = temp;
                right--;
            }else{
                left++;
            }
        }
    }
}